internal/server/recon/recognize.go

```diff
@@ -73,6 +77,174 @@ func RecognizePlaces(
+			}
+			for span := range id2spans[entity.GetInId()] {
+				span2item[span] = &pb.RecognizePlacesResponse_Item{Span: span, Places: places}
+			}
+		}
+
+		// get the list of resolved spans and sort them by more words first. If two
+		// spans have the same number of words, sort alphabetically.
+		spans := []string{}
+		for span := range span2item {
+			spans = append(spans, span)
+		}
+		sort.Slice(spans, func(i, j int) bool {
+			iCount := strings.Count(spans[i], " ")
+			jCount := strings.Count(spans[j], " ")
+			if iCount > jCount {
+				return true
+			} else if jCount > iCount {
+				return false
+			} else {
+				return spans[i] < spans[j]
+			}
+		})
+
+		// iterate through the resolved spans and add the resolved items to the
+		// response
+		queryRespItems := []*pb.RecognizePlacesResponse_Item{{Span: query}}
+		for _, span := range spans {
+			// iterate through the response items to find the item that contains this
+			// curent span. If found, split the response item into multiple items of:
+			// the part of the item before the current span, the item with the current
+			// span and the part of the item after the current span. Do this so that
+			// the spans of the list of items will form the original query.
+			// example:
+			// span: "ab cd"
+			// spans of query resp items: ["testing query ab cd ef g"]
+			// wanted spans of query resp items: ["testing query" "ab cd" "ef g"] and
+			// the item for "ab cd" will have resolved places attached
+			for i, item := range queryRespItems {
+				// if there were Places already detected in this response item, move on
+				// to the next item
+				if len(item.Places) > 0 {
+					continue
+				}
+				// try to split the span of the response item into a list of spans
+				// containing the current span as its own span. if not possible, move on
+				// to the next item.
+				newSpans := getNewSpanList(item.Span, span)
+				if len(newSpans) == 0 {
+					continue
+				}
+				// replace the current response item with a new list of items containing
+				// the new resolved entity
+				newItems := []*pb.RecognizePlacesResponse_Item{}
+				for _, newSpan := range newSpans {
+					trimmedSpan := strings.TrimSpace(newSpan)
+					if len(trimmedSpan) < 1 {
+						continue
+					}
+					if trimmedSpan == span {
+						newItems = append(newItems, span2item[span])
+					} else {
+						newItems = append(newItems, &pb.RecognizePlacesResponse_Item{Span: trimmedSpan})
+					}
+				}
+				newItems = append(newItems, queryRespItems[i+1:]...)
+				queryRespItems = append(queryRespItems[:i], newItems...)
+			}
+		}
+		resp.QueryItems[query] = &pb.RecognizePlacesResponse_Items{Items: queryRespItems}
+	}
+	return resp, nil
+}
+
+// Takes a span and splits it by another span into a list of 3 spans. Example:
+// origSpan: "ab cd ef g", splitSpan: "cd ef", result: ["ab", "cd ef", "g"]
+func getNewSpanList(origSpan string, splitSpan string) []string {
```

Comment:
What if the same span appears multiple times in the query?

Reply:
realized that I did not handle this, updated the function to return a list of non-span part, span part, non-span part, etc.